Expand README description (#61)
paladin-bladesmith/stake-program: README.md

<a href="https://www.npmjs.com/package/@paladin/stake"><img src="https://img.shields.io/npm/v/%40paladin%2Fstake?logo=npm&color=377CC0" /></a>
<a href="https://crates.io/crates/paladin-stake"><img src="https://img.shields.io/crates/v/paladin-stake?logo=rust" /></a>

The Paladin Stake program manages the delegation of tokens to a particular validator of the system. This allows stakers to earn additional shares of rewards proportional to their share of staked tokens and to participate in governance.

## Overview

> A `Config` account essentailly represents the staking system and it is the first account that needs to be created. Once a `Config` account is created, `ValidatorStake` accounts can be added to the system. A `ValidatorStake` represents a validator of the network (`VoteState` account). Anybody can create the stake account for a validator. For new accounts, the authority is initialized to the validator vote account's withdraw authority.

Reviewer: nit: is this maybe clearer?

```suggestion
A `Config` account essentailly represents the staking system and it is the first account that needs to be created. Once a `Config` account is created, `ValidatorStake` accounts can be added to the system. A `ValidatorStake` represents a validator of the network (coupled to a `VoteState` account in the native Solana vote program). Anybody can create the stake account for a validator. For new accounts, the authority is initialized to the validator vote account's withdraw authority.
```

> After `ValidatorStake` accounts are added to the system, `SolStakerStake` accounts can be created. These accounts represent individual SOL stakers (`StakeState` accounts). Similarly to `ValidatorStake`, anybody can create the stake account for a SOL staker. For new accounts, the authority is initialized to the stake state account's withdrawer. When stake is added to a `SolStakerStake`, it will update the total amount of staked tokens both on the correspodent `ValidatorStake` and `Config`. Therefore each `ValidatorStake` tracks the amount of staked tokens that its individual stakers are currently staking, while the `Config` tracks the total of staked tokens on the system. The system also keeps track of the SOL amount staked on the network to determine the share of the rewards and enforce that the proportion of tokens and SOL staked are within the expected limits.

Reviewer: nit: "stake state" isn't very clear
```suggestion
After `ValidatorStake` accounts are added to the system, `SolStakerStake` accounts can be created. These accounts represent individual SOL stakers (coupled to `StakeState` accounts in the native Solana stake program). Similarly to `ValidatorStake`, anybody can create the stake account for a SOL staker. For new accounts, the authority is initialized to the SOL stake account's withdrawer. When stake is added to a `SolStakerStake`, it will update the total amount of staked tokens both on the corresponding `ValidatorStake` and `Config`. Therefore each `ValidatorStake` tracks the amount of staked tokens that its individual stakers are currently staking, while the `Config` tracks the total of staked tokens on the system. The system also keeps track of the SOL amount staked on the network to determine the share of the rewards and enforce that the proportion of tokens and SOL staked are within the expected limits.
```

Staking rewards are paid directly to the program via the `DistributeRewards` instruction while holder rewards are accumulated on the Stake program's `vault` token account. For both cases, the program offer instructions for stakers to harvest their rewards.

> [!IMPORTANT]

> > There can be only one SOL staker stake account per stake state and config account, since the stake state is part of the SOL staker stake account seeds. Similarly, there can be only one validator stake account since since the vote state is part of the validator stake account seeds.

Reviewer: some nits:
```suggestion
> There can be only one SOL staker stake account per SOL stake account and config account, since the SOL stake account is part of the SOL staker stake account seeds. Similarly, there can be only one validator stake account since the vote account is part of the validator stake account seeds.
```

The total amount of SOL staked on a validator is used to determine that maximum amount of tokens that the validator is allowed to stake &mdash; currently the limit is given by `1.3 * SOL amount staked`.

## 📋 Instructions

- [`DeactivateStake`](#deactivatestake)
- [`DistributeRewards`](#distributerewards)
- [`InactivateSolStakerStake`](#inactivatesolstakerstake)
- [`InactivateValidatorStake`](#inactivatevalidatorstake)
- [`InitializeConfig`](#initializeconfig)
- [`InitializeSolStakerStake`](#initializesolstakerstake)
- [`InitializeValidatorStake`](#initializevalidatorstake)
- [`HarvestHolderRewards`](#harvestholderrewards)
- [`HarvestSolStakerRewards`](#harvestsolstakerrewards)
- [`HarvestSyncRewards`](#harvestsyncrewards)
- [`HarvestValidatorRewards`](#harvestvalidatorrewards)
- [`SetAuthority`](#setauthority)
- [`SlashSolStakerStake`](#slashsolstakerstake)
- [`SlashValidatorStake`](#slashvalidatorstake)
- [`SolStakerStakeTokens`](#solstakerstaketokens)
- [`SyncSolStake`](#syncsolstake)
- [`UpdateConfig`](#updateconfig)
- [`ValidatorStakeTokens`](#validatorstaketokens)
- [`WithdrawInactiveStake`](#withdrawinactivestake)

### `DeactivateStake`

> Deactivate staked tokens for a stake delegation, either `ValidatorStake` or `SolStakerStake`. Only one deactivation may be in-flight at once, so if this is called with an active deactivation, it will succeed, but reset the amount and timestamp.

Reviewer: Is it worth adding that deactivating 0 tokens will cancel all deactivations?

Author: Good call, definitely worth mentioning.

### `InitializeSolStakerStake`

Initializes `SolStakerStake` account data for a SOL staker. This instruction can be used multiple times to add a new staker to the stake system. Stakers are uniquely identified by their `StakeState`, i.e., there is only one `SolStakerStake` account for each (`StakeState`, `Config`) pair.

The `SolStakerStake` serves the purpose of managing the stake amount of an individual staker, tracking the SOL amount staked on the network to determine the allowed limit of tokens staked.

> [!NOTE]
> Anybody can create the stake account for a SOL staker. For new accounts, the authority is initialized to the stake state account's withdrawer.

### `InitializeValidatorStake`

Initializes `ValidatorStake` account data for a validator. This instruction can be used multiple times to add validators to the stake system. Validators are uniquely identified by their `VoteState`, i.e., there is only one `ValidatorStake` account for each (`VoteState`, `Config`) pair.

> The `ValidatorStake` serves two purposes on the staking system: (1) it allows individual staker (`SolStakerStake` account) to stake tokens on the system; and (2) it allows validators to stake tokens on the system. Each validator tracks the SOL amount staked on the network of its stakers, which in turn determines the amount of tokens that a validator and its stakers are allows to stake.

Reviewer: nit typo
```suggestion
The `ValidatorStake` serves two purposes on the staking system: (1) it allows individual staker (`SolStakerStake` account) to stake tokens on the system; and (2) it allows validators to stake tokens on the system. Each validator tracks the SOL amount staked on the network of its stakers, which in turn determines the amount of tokens that a validator and its stakers are allowed to stake.
```

> [!NOTE]
>  Anybody can create the stake account for a validator. For new accounts, the authority is initialized to the validator vote account's withdraw authority.

### `HarvestHolderRewards`

> Harvests holder SOL rewards earned by the given stake account. This instruction support claiming rewards for both `ValidatorStake` and `SolStakerStake` accounts.

Reviewer: ```suggestion
Harvests holder SOL rewards earned by the given stake account. This instruction supports claiming rewards for both `ValidatorStake` and `SolStakerStake` accounts.
```

### `HarvestSolStakerRewards`

> Harvests stake SOL rewards earned by the given SOL staker stake account.

Reviewer: ```suggestion
Harvests staker SOL rewards earned by the given SOL staker stake account.
```

### `HarvestSyncRewards`

Harvest the rewards from syncing the SOL stake balance with a validator and SOL staker stake accounts.

> The staking system requires the SOL staked amount to be up to date with the `StakeState` network delegation amount. This instruction is used to sync their amounts, rewarding the address that executes the instrution a reward.

Reviewer: ```suggestion
The staking system requires the SOL staked amount to be up to date with the `StakeState` network delegation amount. This instruction is used to sync their amounts, rewarding the address that executes the instruction.
```

> [!NOTE]
> This is a permissionless instruction, anybody can sync the balance of a SOL stake account. Rewards are only paid when the balances are out-of-sync.

### `HarvestValidatorRewards`

> Harvests stake SOL rewards earned by the given validator stake account.

Reviewer: ```suggestion
Harvests staker SOL rewards earned by the given validator stake account.
```
